File: core/src/main/java/dev/bee/kanjianki/core/AutoSyncSchedulePolicy.java
package dev.bee.kanjianki.core;

import java.util.Calendar;

public final class AutoSyncSchedulePolicy {
    public static final long MIN_DELAY_MILLIS = 10_000L;
    public static final long DEADLINE_WINDOW_MILLIS = 6L * 60L * 60L * 1000L;

    private AutoSyncSchedulePolicy() {
    }

    public static SchedulePlan plan(
            boolean enabled,
            int hour,
            int minute,
            long nowMillis,
            boolean alreadySyncedToday
    ) {
        if (!enabled) {
            return SchedulePlan.disabled();
        }
        long triggerAt = nextTriggerMillis(hour, minute, nowMillis, alreadySyncedToday);
        long minimumLatency = Math.max(MIN_DELAY_MILLIS, triggerAt - nowMillis);
        return planWithLatency(triggerAt, minimumLatency);
    }

    public static SchedulePlan planAt(long triggerAtMillis, long nowMillis) {
        return planWithLatency(triggerAtMillis, Math.max(MIN_DELAY_MILLIS, triggerAtMillis - nowMillis));
    }

    public static long nextTriggerMillis(int hour, int minute, long nowMillis) {
        return nextTriggerMillis(hour, minute, nowMillis, false);
    }

    public static long nextTriggerMillis(int hour, int minute, long nowMillis, boolean alreadySyncedToday) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(nowMillis);
        calendar.set(Calendar.HOUR_OF_DAY, hour);
        calendar.set(Calendar.MINUTE, minute);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        long trigger = calendar.getTimeInMillis();
        if (trigger <= nowMillis || alreadySyncedToday) {
            calendar.add(Calendar.DAY_OF_YEAR, 1);
            trigger = calendar.getTimeInMillis();
        }
        return trigger;
    }

    public static long localDayStart(long nowMillis) {
        return LocalDayPolicy.localDayStart(nowMillis);
    }

    private static SchedulePlan planWithLatency(long triggerAtMillis, long minimumLatencyMillis) {
        return new SchedulePlan(
                true,
                triggerAtMillis,
                minimumLatencyMillis,
                minimumLatencyMillis + DEADLINE_WINDOW_MILLIS
        );
    }

    public record SchedulePlan(
            boolean enabled,
            long triggerAtMillis,
            long minimumLatencyMillis,
            long overrideDeadlineMillis
    ) {
        static SchedulePlan disabled() {
            return new SchedulePlan(false, 0L, 0L, 0L);
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/AutoSyncSchedulePolicy.kt
package dev.bee.kanjianki.core

import java.util.Calendar
import kotlin.math.max

object AutoSyncSchedulePolicy {
    const val MIN_DELAY_MILLIS: Long = 10_000L
    const val DEADLINE_WINDOW_MILLIS: Long = 6L * 60L * 60L * 1000L

    @JvmStatic
    fun plan(
        enabled: Boolean,
        hour: Int,
        minute: Int,
        nowMillis: Long,
        alreadySyncedToday: Boolean,
    ): SchedulePlan {
        if (!enabled) {
            return SchedulePlan.disabled()
        }
        val triggerAt = nextTriggerMillis(hour, minute, nowMillis, alreadySyncedToday)
        val minimumLatency = max(MIN_DELAY_MILLIS, triggerAt - nowMillis)
        return planWithLatency(triggerAt, minimumLatency)
    }

    @JvmStatic
    fun planAt(triggerAtMillis: Long, nowMillis: Long): SchedulePlan {
        return planWithLatency(triggerAtMillis, max(MIN_DELAY_MILLIS, triggerAtMillis - nowMillis))
    }

    @JvmStatic
    fun nextTriggerMillis(hour: Int, minute: Int, nowMillis: Long): Long {
        return nextTriggerMillis(hour, minute, nowMillis, false)
    }

    @JvmStatic
    fun nextTriggerMillis(hour: Int, minute: Int, nowMillis: Long, alreadySyncedToday: Boolean): Long {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = nowMillis
        calendar.set(Calendar.HOUR_OF_DAY, hour)
        calendar.set(Calendar.MINUTE, minute)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        var trigger = calendar.timeInMillis
        if (trigger <= nowMillis || alreadySyncedToday) {
            calendar.add(Calendar.DAY_OF_YEAR, 1)
            trigger = calendar.timeInMillis
        }
        return trigger
    }

    @JvmStatic
    fun localDayStart(nowMillis: Long): Long {
        return LocalDayPolicy.localDayStart(nowMillis)
    }

    private fun planWithLatency(triggerAtMillis: Long, minimumLatencyMillis: Long): SchedulePlan {
        return SchedulePlan(
            true,
            triggerAtMillis,
            minimumLatencyMillis,
            minimumLatencyMillis + DEADLINE_WINDOW_MILLIS,
        )
    }

    @JvmRecord
    data class SchedulePlan(
        val enabled: Boolean,
        val triggerAtMillis: Long,
        val minimumLatencyMillis: Long,
        val overrideDeadlineMillis: Long,
    ) {
        override fun toString(): String {
            return "SchedulePlan[enabled=$enabled, triggerAtMillis=$triggerAtMillis, " +
                "minimumLatencyMillis=$minimumLatencyMillis, overrideDeadlineMillis=$overrideDeadlineMillis]"
        }

        companion object {
            @JvmStatic
            fun disabled(): SchedulePlan {
                return SchedulePlan(false, 0L, 0L, 0L)
            }
        }
    }
}
